### PR TITLE
Fix paginating /relations with a live token

### LINKED ISSUE #14830: Internal server error when calling /relations with a token from /sync
```
2023-01-12 16:41:20,025 - synapse.http.server - 124 - ERROR - GET-832933- Failed handle request via 'RelationPaginationServlet': <XForwardedForRequest at 0x7f5a7a2c98c0 method='GET' uri='/_matrix/client/v1/rooms/!test%3Amatrix.org/relations/%24foo/m.thread?dir=b&limit=20&from=sXXXX_1_YYY_ZZZ_AAA_BBB_CCC_DDD_0' clientproto='HTTP/1.1' site='8008'>
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/synapse/util/caches/descriptors.py", line 243, in _wrapped
    ret = cache.get(cache_key, callback=invalidate_callback)
  File "/usr/local/lib/python3.9/site-packages/synapse/util/caches/deferred_cache.py", line 180, in get
    raise KeyError()
KeyError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/synapse/http/server.py", line 307, in _async_render_wrapper
    callback_return = await self._async_render(request)
  File "/usr/local/lib/python3.9/site-packages/synapse/http/server.py", line 513, in _async_render
    callback_return = await raw_callback_return
  File "/usr/local/lib/python3.9/site-packages/synapse/rest/client/relations.py", line 71, in on_GET
    result = await self._relations_handler.get_relations(
  File "/usr/local/lib/python3.9/site-packages/synapse/handlers/relations.py", line 125, in get_relations
    related_events, next_token = await self._main_store.get_relations_for_event(
  File "/usr/local/lib/python3.9/site-packages/twisted/internet/defer.py", line 1697, in _inlineCallbacks
    result = context.run(gen.send, result)
  File "/usr/local/lib/python3.9/site-packages/synapse/storage/databases/main/relations.py", line 213, in get_relations_for_event
    from_token=from_token.room_key.as_historical_tuple()
  File "/usr/local/lib/python3.9/site-packages/synapse/types.py", line 569, in as_historical_tuple
    raise Exception(
Exception: Cannot call `RoomStreamToken.as_historical_tuple` on live token
```

https://sentry.tools.element.io/organizations/element/issues/58936/

### PR DESCRIPTION
**Pull request by clokep**
The `/relations` endpoint was not properly handle "live tokens" (i.e sync tokens), to do this properly we abstract the code that `/messages` has and re-use it.

Tests at matrix-org/complement#586.

Reviewable commit-by-commit.

Fixes #14830

Related to #13862

### CODE REVIEW

**Review by DMRobertson — APPROVED** [review]
Many thanks for this---despaghettification much appreciated.

File: synapse/storage/databases/main/stream.py [hunk 1]
```diff
@@ -170,6 +170,74 @@ def generate_pagination_where_clause(
...
 
 
+def generate_pagination_bounds(
+    direction: str, from_token: RoomStreamToken, to_token: Optional[RoomStreamToken]
+) -> Tuple[str, Tuple[Optional[int], int], Optional[Tuple[Optional[int], int]]]:
+    """
+    Generate a start and end point for this page of events.
+
+    Args:
+        direction: Whether pagination is going forwards or backwards. One of "f" or "b".
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/stream.py:184]: I think we can make it an enum even, I'll take a look as a follow-up.

[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/storage/databases/main/stream.py:184]: SGTM either way. But yeah, definitely one for a follow-up.

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/stream.py:184]: See #14927.

**Pull request merged**